File: scripts/enrich-gyms.ts
// Enrich gym profiles by scraping their websites and extracting data via Claude Haiku.
//
// Usage:
//   npx tsx scripts/enrich-gyms.ts                     # dry-run (preview only)
//   npx tsx scripts/enrich-gyms.ts --apply             # actually write to DynamoDB
//   npx tsx scripts/enrich-gyms.ts --apply --limit 5   # limit to first 5 gyms
//   npx tsx scripts/enrich-gyms.ts --skip-copy         # skip delete/copy, just enrich
//   npx tsx scripts/enrich-gyms.ts --state WA          # only enrich gyms in a specific state
//
// Steps:
//   1. Delete all gyms in staging table
//   2. Copy all gyms from prod table
//   3. For each gym with a website (where adminEdited is null/false):
//      - Fetch website HTML, convert to text
//      - Send to Claude Haiku for extraction
//      - Update DynamoDB with extracted fields

import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  ScanCommand,
  BatchWriteCommand,
  UpdateCommand,
} from "@aws-sdk/lib-dynamodb";
import { SSMClient, GetParametersCommand } from "@aws-sdk/client-ssm";

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------
const STAGING_TABLE = "Gym-qanzfeewlfeklctnhoskryahti-NONE";
const PROD_TABLE = "Gym-xofowsmrxvebxmdjijmijtz5bq-NONE";
const REGION = "ap-southeast-2";
const APP_ID = "d36uz2q25gygnh";

const ddb = DynamoDBDocumentClient.from(new DynamoDBClient({ region: REGION }));
const args = process.argv.slice(2);
const APPLY = args.includes("--apply");
const SKIP_COPY = args.includes("--skip-copy");
const limitIdx = args.indexOf("--limit");
const LIMIT = limitIdx >= 0 ? parseInt(args[limitIdx + 1]) : Infinity;
const stateIdx = args.indexOf("--state");
const STATE_FILTER = stateIdx >= 0 ? args[stateIdx + 1]?.toUpperCase() : "";

// ---------------------------------------------------------------------------
// SSM -> Anthropic API key
// ---------------------------------------------------------------------------
let anthropicKey = "";

async function getAnthropicKey(): Promise<string> {
  if (anthropicKey) return anthropicKey;
  if (process.env.ANTHROPIC_API_KEY) {
    anthropicKey = process.env.ANTHROPIC_API_KEY;
    return anthropicKey;
  }
  const client = new SSMClient({ region: REGION });
  const result = await client.send(
    new GetParametersCommand({
      Names: [
        `/amplify/shared/${APP_ID}/ANTHROPIC_API_KEY`,
        `/amplify/${APP_ID}/staging/ANTHROPIC_API_KEY`,
      ],
      WithDecryption: true,
    })
  );
  for (const param of result.Parameters ?? []) {
    if (param.Value) { anthropicKey = param.Value; break; }
  }
  if (!anthropicKey) throw new Error("Could not retrieve ANTHROPIC_API_KEY from SSM or env");
  return anthropicKey;
}

// ---------------------------------------------------------------------------
// HTML -> text
// ---------------------------------------------------------------------------
function htmlToText(html: string): string {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, "")
    .replace(/<style[\s\S]*?<\/style>/gi, "")
    .replace(/<noscript[\s\S]*?<\/noscript>/gi, "")
    .replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, "$2 ($1)")
    .replace(/<\/(p|div|h[1-6]|li|tr|br\s*\/?)>/gi, "\n")
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<[^>]+>/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, " ")
    .replace(/[ \t]+/g, " ")
    .replace(/\n\s*\n/g, "\n")
    .trim();
}

// ---------------------------------------------------------------------------
// Step 1: Delete all staging gyms
// ---------------------------------------------------------------------------
async function deleteAllStaging() {
  console.log("\n=== Step 1: Deleting all staging gyms ===");
  const items: Record<string, unknown>[] = [];
  let lastKey: Record<string, unknown> | undefined;
  do {
    const res = await ddb.send(new ScanCommand({
      TableName: STAGING_TABLE,
      ProjectionExpression: "id",
      ExclusiveStartKey: lastKey,
    }));
    items.push(...(res.Items ?? []));
    lastKey = res.LastEvaluatedKey;
  } while (lastKey);

  console.log(`  Found ${items.length} staging gyms to delete`);
  if (!APPLY) { console.log("  [DRY RUN] Skipping delete"); return; }

  const BATCH = 25;
  for (let i = 0; i < items.length; i += BATCH) {
    const batch = items.slice(i, i + BATCH);
    await ddb.send(new BatchWriteCommand({
      RequestItems: {
        [STAGING_TABLE]: batch.map((item) => ({
          DeleteRequest: { Key: { id: item.id } },
        })),
      },
    }));
  }
  console.log(`  Deleted ${items.length} staging gyms`);
}

// ---------------------------------------------------------------------------
// Step 2: Copy prod -> staging
// ---------------------------------------------------------------------------
async function copyProdToStaging(): Promise<Record<string, unknown>[]> {
  console.log("\n=== Step 2: Copying prod gyms to staging ===");
  const items: Record<string, unknown>[] = [];
  let lastKey: Record<string, unknown> | undefined;
  do {
    const res = await ddb.send(new ScanCommand({
      TableName: PROD_TABLE,
      ExclusiveStartKey: lastKey,
    }));
    items.push(...(res.Items ?? []));
    lastKey = res.LastEvaluatedKey;
  } while (lastKey);

  // Fix empty GSI key fields — DynamoDB rejects empty strings on GSI keys
  let fixed = 0;
  for (const item of items) {
    if (item.addressPostcode === "") { delete item.addressPostcode; fixed++; }
    if (item.ownerId === "") { delete item.ownerId; fixed++; }
  }
  console.log(`  Found ${items.length} prod gyms${fixed > 0 ? ` (fixed ${fixed} empty GSI keys)` : ""}`);
  if (!APPLY) { console.log("  [DRY RUN] Skipping copy"); return items; }

  const BATCH = 25;
  let written = 0;
  for (let i = 0; i < items.length; i += BATCH) {
    const batch = items.slice(i, i + BATCH);
    await ddb.send(new BatchWriteCommand({
      RequestItems: {
        [STAGING_TABLE]: batch.map((item) => ({
          PutRequest: { Item: item },
        })),
      },
    }));
    written += batch.length;
    if (written % 500 === 0) console.log(`  ${written} / ${items.length} copied...`);
  }
  console.log(`  Copied ${written} gyms to staging`);
  return items;
}

// ---------------------------------------------------------------------------
// Step 3: Enrich gyms
// ---------------------------------------------------------------------------
const SYSTEM_PROMPT = `You are analyzing a gym/fitness centre website. Extract as much useful profile information as possible from the page content.

Return a JSON object with ONLY the fields you can confidently extract. Omit any field you cannot find or are unsure about. Do NOT guess or make up data.

Fields to look for:
- "phone": phone number (Australian format preferred)
- "email": email address
- "instagram": Instagram profile URL (full URL)
- "facebook": Facebook page URL (full URL)
- "bookingUrl": online booking URL (Mindbody, Glofox, etc.)
- "amenities": array of amenities from this list ONLY: ["pool", "spa", "sauna", "free weights", "cardio", "group classes", "boxing/mma", "yoga/pilates", "parking", "showers", "lockers", "childcare", "cafe", "24/7 access", "personal training"]
- "specialties": array of specialties/programs (e.g. "Hyrox", "F45", "Olympic Lifting", "Powerlifting", "Rehab", "Seniors Fitness", "CrossFit", "HIIT", "Reformer Pilates", "Spin/Cycle", "Martial Arts", "Muay Thai", "Boxing", "Kickboxing", "Strength & Conditioning", "Functional Training", "Circuit Training")
- "description": write a compelling, SEO-optimized description for this gym listing on mynextgym.com.au. 100-200 words, plain text only (no HTML, no markdown, no headings). Use Australian English spelling. Emphasise the gym's amenities, location, and what makes it stand out. Naturally include the suburb name for local SEO. The tone should be professional but approachable - as if the gym owner is talking to a potential member. Do NOT include the gym name at the start (it's already shown as the page title). Do NOT use bullet points or lists. Write flowing paragraphs only. Do NOT copy text verbatim from the website - rewrite it in your own words
- "pricePerWeek": weekly membership price as a number (calculate from weekly/fortnightly/monthly if needed)
- "hours": object with keys monday-sunday, values as opening hours strings (e.g. "5:00am - 9:00pm"). Only include days where hours are clearly stated on the website.
- "memberOffers": array of member offer tags from this list ONLY: ["no contract", "contract", "new member trial", "referral scheme", "multiple location access", "gym or community app", "casual classes"]

Return ONLY valid JSON. No markdown, no explanation, no code blocks.`;

async function fetchWebsite(url: string): Promise<string | null> {
  try {
    const parsedUrl = new URL(url.startsWith("http") ? url : `https://${url}`);
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10000);

    const response = await fetch(parsedUrl.toString(), {
      signal: controller.signal,
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; mynextgym.com.au bot; +https://www.mynextgym.com.au)",
        Accept: "text/html,application/xhtml+xml",
      },
    });
    clearTimeout(timeout);

    if (!response.ok) return null;
    const html = await response.text();
    let text = htmlToText(html);
    if (text.length > 8000) text = text.slice(0, 8000);
    if (text.length < 50) return null;
    return text;
  } catch {
    return null;
  }
}

async function extractWithHaiku(
  apiKey: string,
  websiteUrl: string,
  pageText: string
): Promise<Record<string, unknown> | null> {
  try {
    const response = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify({
        model: "claude-haiku-4-5-20251001",
        max_tokens: 2048,
        system: SYSTEM_PROMPT,
        messages: [{ role: "user", content: `Website URL: ${websiteUrl}\n\nPage content:\n${pageText}` }],
      }),
    });

    if (!response.ok) {
      const err = await response.text();
      console.error(`    Haiku error ${response.status}: ${err.slice(0, 200)}`);
      return null;
    }

    const data = await response.json();
    const raw = data.content?.[0]?.text ?? "{}";
    const cleaned = raw.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/i, "").trim();
    return JSON.parse(cleaned);
  } catch (err) {
    console.error(`    Haiku parse error:`, err);
    return null;
  }
}

async function enrichGyms(gyms: Record<string, unknown>[]) {
  console.log("\n=== Step 3: Enriching gyms ===");

  // Filter: has website, adminEdited is null/false, optional state filter
  const candidates = gyms.filter(
    (g) =>
      g.website &&
      String(g.website).length > 5 &&
      !g.adminEdited &&
      (!STATE_FILTER || String(g.addressState || "").toUpperCase() === STATE_FILTER)
  );

  const toProcess = candidates.slice(0, LIMIT);
  const stateLabel = STATE_FILTER ? ` [${STATE_FILTER} only]` : "";
  console.log(`  ${candidates.length} candidates${stateLabel} (adminEdited=null/false with website), processing ${toProcess.length}\n`);

  if (!APPLY) {
    console.log("DRY-RUN — skipping API calls. Run with --apply to enrich.\n");
    return;
  }

  const apiKey = await getAnthropicKey();
  console.log("  API key retrieved");

  let updated = 0;
  let skipped = 0;
  let failed = 0;

  for (let i = 0; i < toProcess.length; i++) {
    const gym = toProcess[i];
    const name = String(gym.name ?? "");
    const website = String(gym.website ?? "");
    const id = String(gym.id);

    console.log(`  [${i + 1}/${toProcess.length}] ${name} -- ${website}`);

    // Fetch website
    const pageText = await fetchWebsite(website);
    if (!pageText) {
      console.log("    ! Could not fetch website, skipping");
      skipped++;
      continue;
    }

    // Extract via Haiku
    const extracted = await extractWithHaiku(apiKey, website, pageText);
    if (!extracted) {
      console.log("    ! Extraction failed, skipping");
      failed++;
      continue;
    }

    // Build update expression
    const setExprs: string[] = ["updatedAt = :now"];
    const values: Record<string, unknown> = { ":now": new Date().toISOString() };
    const updatedFields: string[] = [];

    // Description
    if (extracted.description && typeof extracted.description === "string") {
      setExprs.push("description = :desc");
      values[":desc"] = extracted.description;
      updatedFields.push("description");
    }

    // Price + pricing notes
    if (extracted.pricePerWeek && typeof extracted.pricePerWeek === "number") {
      setExprs.push("pricePerWeek = :price");
      values[":price"] = extracted.pricePerWeek;
      setExprs.push("pricingNotes = :pnotes");
      values[":pnotes"] = "Verified by AI";
      updatedFields.push("pricePerWeek");
    }

    // Phone
    if (extracted.phone && typeof extracted.phone === "string") {
      setExprs.push("phone = :phone");
      values[":phone"] = extracted.phone;
      updatedFields.push("phone");
    }

    // Email
    if (extracted.email && typeof extracted.email === "string") {
      setExprs.push("email = :email");
      values[":email"] = extracted.email;
      updatedFields.push("email");
    }

    // Instagram
    if (extracted.instagram && typeof extracted.instagram === "string") {
      setExprs.push("instagram = :ig");
      values[":ig"] = extracted.instagram;
      updatedFields.push("instagram");
    }

    // Facebook
    if (extracted.facebook && typeof extracted.facebook === "string") {
      setExprs.push("facebook = :fb");
      values[":fb"] = extracted.facebook;
      updatedFields.push("facebook");
    }

    // Booking URL
    if (extracted.bookingUrl && typeof extracted.bookingUrl === "string") {
      setExprs.push("bookingUrl = :booking");
      values[":booking"] = extracted.bookingUrl;
      updatedFields.push("bookingUrl");
    }

    // Amenities + notes
    if (Array.isArray(extracted.amenities) && extracted.amenities.length > 0) {
      setExprs.push("amenities = :amenities");
      values[":amenities"] = extracted.amenities;
      setExprs.push("amenitiesNotes = :anotes");
      values[":anotes"] = "Verified by AI";
      updatedFields.push(`amenities(${extracted.amenities.length})`);
    }

    // Specialties
    if (Array.isArray(extracted.specialties) && extracted.specialties.length > 0) {
      setExprs.push("specialties = :specs");
      values[":specs"] = extracted.specialties;
      updatedFields.push(`specialties(${extracted.specialties.length})`);
    }

    // Member offers
    if (Array.isArray(extracted.memberOffers) && extracted.memberOffers.length > 0) {
      setExprs.push("memberOffers = :offers");
      values[":offers"] = extracted.memberOffers;
      updatedFields.push(`memberOffers(${extracted.memberOffers.length})`);
    }

    // Hours -- only include days with real values
    const hours = extracted.hours as Record<string, string> | undefined;
    if (hours && typeof hours === "object") {
      const dayMap: Record<string, string> = {
        monday: "hoursMonday", tuesday: "hoursTuesday", wednesday: "hoursWednesday",
        thursday: "hoursThursday", friday: "hoursFriday", saturday: "hoursSaturday",
        sunday: "hoursSunday",
      };
      let hoursCount = 0;
      for (const [day, attr] of Object.entries(dayMap)) {
        const val = hours[day];
        if (val && typeof val === "string" && val.length > 2 &&
            !val.toLowerCase().includes("unavailable") &&
            !val.toLowerCase().includes("unknown") &&
            !val.toLowerCase().includes("closed") &&
            !val.toLowerCase().includes("n/a")) {
          const placeholder = `:h${day.slice(0, 3)}`;
          setExprs.push(`${attr} = ${placeholder}`);
          values[placeholder] = val;
          hoursCount++;
        }
      }
      if (hoursCount > 0) updatedFields.push(`hours(${hoursCount} days)`);
    }

    if (updatedFields.length === 0) {
      console.log("    ! No fields extracted, skipping");
      skipped++;
      continue;
    }

    console.log(`    > Extracted: ${updatedFields.join(", ")}`);

    if (APPLY) {
      await ddb.send(new UpdateCommand({
        TableName: STAGING_TABLE,
        Key: { id },
        UpdateExpression: `SET ${setExprs.join(", ")}`,
        ExpressionAttributeValues: values,
      }));
      console.log("    > Written to DynamoDB");
    } else {
      console.log("    [DRY RUN] Would update DynamoDB");
    }

    updated++;

    // Small delay to avoid rate limits
    await new Promise((r) => setTimeout(r, 500));
  }

  console.log(`\n=== Summary ===`);
  console.log(`  Updated: ${updated}`);
  console.log(`  Skipped: ${skipped}`);
  console.log(`  Failed: ${failed}`);
  console.log(`  Mode: ${APPLY ? "APPLIED" : "DRY RUN"}`);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main() {
  console.log(`Mode: ${APPLY ? "APPLY (writing to DynamoDB)" : "DRY RUN (preview only)"}`);
  console.log(`Limit: ${LIMIT === Infinity ? "all" : LIMIT}`);

  if (!SKIP_COPY) {
    await deleteAllStaging();
    await copyProdToStaging();
  }

  // Scan staging table for enrichment
  const gyms: Record<string, unknown>[] = [];
  let lastKey: Record<string, unknown> | undefined;
  do {
    const res = await ddb.send(new ScanCommand({
      TableName: SKIP_COPY ? STAGING_TABLE : STAGING_TABLE,
      ExclusiveStartKey: lastKey,
    }));
    gyms.push(...(res.Items ?? []));
    lastKey = res.LastEvaluatedKey;
  } while (lastKey);

  await enrichGyms(gyms);
}

main().catch((err) => {
  console.error("Script failed:", err);
  process.exit(1);
});

File: scripts/enrich_gyms.py
#!/usr/bin/env python3
"""
Gym amenities enrichment script.

Reads data/gyms_dynamo.csv, fetches each gym's website via Claude + web tools,
extracts amenities / pricing / member offers, and writes to data/gyms_enriched.csv.

Usage:
    pip install anthropic
    export ANTHROPIC_API_KEY=sk-ant-...
    python scripts/enrich_gyms.py

Resume:  re-running automatically skips already-processed gym IDs.
Filter:  python scripts/enrich_gyms.py --state VIC   (process one state only)
Limit:   python scripts/enrich_gyms.py --limit 50    (process first N gyms)
"""

import argparse
import csv
import json
import os
import sys
import time
from pathlib import Path

# ── Load .env.local ───────────────────────────────────────────────────────────
_env_file = Path(__file__).parent.parent / ".env.local"
if _env_file.exists():
    for _line in _env_file.read_text().splitlines():
        _line = _line.strip()
        if _line and not _line.startswith("#") and "=" in _line:
            _k, _, _v = _line.partition("=")
            os.environ.setdefault(_k.strip(), _v.strip().strip('"').strip("'"))

import anthropic

# ── Config ────────────────────────────────────────────────────────────────────
INPUT_CSV  = Path("data/gyms_dynamo.csv")
OUTPUT_CSV = Path("data/gyms_enriched.csv")

MODEL          = "claude-opus-4-6"
MAX_RETRIES    = 3
RATE_LIMIT_S   = 2.0   # seconds between calls (adjust for your API tier)
MAX_CONTINUATIONS = 6  # max pause_turn resume loops per gym

OUTPUT_FIELDS = [
    "id", "name", "addressState", "website",
    "status",                # success | failed | no_website | parse_error
    "min_weekly_price_aud",
    "amenities",             # JSON string
    "member_offers",         # JSON string
    "confidence",            # high | medium | low | no_data
    "error",
]

SYSTEM_PROMPT = """\
You are a gym data enrichment assistant. Your job is to research gym websites \
and extract structured membership information.

Use web_fetch on the provided URL (and follow links to pricing/membership pages if needed). \
Use web_search only if web_fetch fails or yields no useful content.

Your FINAL response must be ONLY valid JSON — no preamble, no explanation, no markdown fences.
Use exactly this schema (all fields required, use null if unknown):

{
  "min_weekly_price_aud": <number or null>,
  "confidence": "high" | "medium" | "low" | "no_data",
  "amenities": {
    "pool": bool,
    "spa": bool,
    "sauna": bool,
    "free_weights": bool,
    "cardio": bool,
    "group_classes": bool,
    "boxing_mma": bool,
    "yoga_pilates": bool,
    "parking": bool,
    "showers": bool,
    "lockers": bool,
    "childcare": bool,
    "cafe": bool,
    "access_247": bool,
    "personal_training": bool
  },
  "member_offers": {
    "no_contract": bool,
    "contract": bool,
    "new_member_trial": bool,
    "referral_scheme": bool,
    "multi_location_access": bool,
    "app": bool
  }
}

confidence levels:
- "high"    = price and most amenities found clearly on the website
- "medium"  = some data found but pricing missing or ambiguous
- "low"     = very little data found, mostly guessing from gym type
- "no_data" = website unreachable or no relevant content found
"""

TOOLS = [
    {"type": "web_search_20260209", "name": "web_search"},
    {"type": "web_fetch_20260209",  "name": "web_fetch"},
]

DEFAULT_RESULT = {
    "min_weekly_price_aud": None,
    "confidence": "no_data",
    "amenities": {
        "pool": False, "spa": False, "sauna": False, "free_weights": False,
        "cardio": False, "group_classes": False, "boxing_mma": False,
        "yoga_pilates": False, "parking": False, "showers": False,
        "lockers": False, "childcare": False, "cafe": False,
        "access_247": False, "personal_training": False,
    },
    "member_offers": {
        "no_contract": False, "contract": False, "new_member_trial": False,
        "referral_scheme": False, "multi_location_access": False, "app": False,
    },
}


# ── Helpers ───────────────────────────────────────────────────────────────────

def load_processed_ids(path: Path) -> set:
    if not path.exists():
        return set()
    with path.open(newline="", encoding="utf-8") as f:
        return {row["id"] for row in csv.DictReader(f)}


def append_row(path: Path, row: dict):
    is_new = not path.exists()
    with path.open("a", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=OUTPUT_FIELDS, extrasaction="ignore")
        if is_new:
            writer.writeheader()
        writer.writerow(row)


def extract_json(text: str) -> dict | None:
    """Extract JSON object from Claude's response text."""
    text = text.strip()
    # Strip markdown fences if present
    if text.startswith("```"):
        lines = text.splitlines()
        text = "\n".join(lines[1:-1] if lines[-1].strip() == "```" else lines[1:])
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        # Try to find a JSON object within the text
        start = text.find("{")
        end   = text.rfind("}") + 1
        if start != -1 and end > start:
            try:
                return json.loads(text[start:end])
            except json.JSONDecodeError:
                pass
    return None


def build_prompt(gym: dict) -> str:
    parts = [f"Gym name: {gym['name']}"]
    if gym.get("website"):
        parts.append(f"Website: {gym['website']}")
    if gym.get("addressSuburb"):
        parts.append(f"Location: {gym['addressSuburb']}, {gym.get('addressState', '')}")
    if gym.get("description"):
        parts.append(f"Description: {gym['description'][:400]}")
    return "\n".join(parts)


def call_claude(client: anthropic.Anthropic, gym: dict) -> tuple[str, dict, str | None]:
    """
    Returns (status, result_dict, error_message).
    status: "success" | "parse_error" | "api_error"
    """
    messages = [{"role": "user", "content": build_prompt(gym)}]
    last_content = None

    for _ in range(MAX_CONTINUATIONS):
        response = client.messages.create(
            model=MODEL,
            max_tokens=2048,
            system=SYSTEM_PROMPT,
            tools=TOOLS,
            messages=messages,
        )

        # Collect text from response
        for block in response.content:
            if hasattr(block, "text"):
                last_content = block.text

        if response.stop_reason == "end_turn":
            break

        if response.stop_reason == "pause_turn":
            # Server-side tools hit iteration limit — re-send to continue
            messages = [
                {"role": "user",      "content": build_prompt(gym)},
                {"role": "assistant", "content": response.content},
            ]
            continue

        # Unexpected stop
        break

    if not last_content:
        return "parse_error", DEFAULT_RESULT, "No text in response"

    parsed = extract_json(last_content)
    if not parsed:
        return "parse_error", DEFAULT_RESULT, f"Could not parse JSON: {last_content[:200]}"

    # Merge with defaults so missing keys don't cause KeyErrors downstream
    result = json.loads(json.dumps(DEFAULT_RESULT))  # deep copy
    result.update({k: v for k, v in parsed.items() if k in result})
    if "amenities" in parsed:
        result["amenities"].update(parsed["amenities"])
    if "member_offers" in parsed:
        result["member_offers"].update(parsed["member_offers"])

    return "success", result, None


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Enrich gym data via Claude web tools")
    parser.add_argument("--state",  help="Filter to a single state (e.g. VIC, QLD)")
    parser.add_argument("--limit",  type=int, help="Max gyms to process this run")
    parser.add_argument("--delay",  type=float, default=RATE_LIMIT_S,
                        help=f"Seconds between API calls (default {RATE_LIMIT_S})")
    args = parser.parse_args()

    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        print("Error: ANTHROPIC_API_KEY environment variable not set", file=sys.stderr)
        sys.exit(1)

    client = anthropic.Anthropic(api_key=api_key)

    # Load input CSV
    with INPUT_CSV.open(newline="", encoding="utf-8") as f:
        all_gyms = list(csv.DictReader(f))

    print(f"Loaded {len(all_gyms):,} gyms from {INPUT_CSV}")

    # Apply state filter
    if args.state:
        all_gyms = [g for g in all_gyms if g.get("addressState", "").upper() == args.state.upper()]
        print(f"Filtered to {len(all_gyms):,} gyms in {args.state.upper()}")

    # Skip already-processed
    processed = load_processed_ids(OUTPUT_CSV)
    remaining = [g for g in all_gyms if g["id"] not in processed]
    print(f"Already processed: {len(processed):,}  |  Remaining: {len(remaining):,}")

    # Apply limit
    if args.limit:
        remaining = remaining[:args.limit]
        print(f"Processing first {len(remaining):,} (--limit {args.limit})")

    if not remaining:
        print("Nothing to do.")
        return

    counts = {"success": 0, "failed": 0, "no_website": 0, "parse_error": 0}

    for i, gym in enumerate(remaining, 1):
        name    = gym.get("name", "?")
        website = gym.get("website", "").strip()
        state   = gym.get("addressState", "")

        label = f"[{i:>{len(str(len(remaining)))}}/{len(remaining)}] {name[:55]:<55} ({state})"
        print(label, end=" ... ", flush=True)

        # No website
        if not website:
            append_row(OUTPUT_CSV, {
                "id": gym["id"], "name": name, "addressState": state,
                "website": "", "status": "no_website",
                "min_weekly_price_aud": "", "amenities": "", "member_offers": "",
                "confidence": "", "error": "",
            })
            counts["no_website"] += 1
            print("SKIP (no website)")
            continue

        # Retry loop
        status = "failed"
        result = DEFAULT_RESULT
        error  = None

        for attempt in range(1, MAX_RETRIES + 1):
            try:
                status, result, error = call_claude(client, gym)
                break
            except anthropic.RateLimitError:
                wait = 30 * attempt
                print(f"rate-limited, waiting {wait}s ...", end=" ", flush=True)
                time.sleep(wait)
                error = "rate_limit"
            except anthropic.APIStatusError as e:
                error = f"{e.status_code}: {str(e)[:100]}"
                if attempt < MAX_RETRIES:
                    time.sleep(5 * attempt)
            except Exception as e:
                error = str(e)[:120]
                break

        if status in ("success", "parse_error"):
            counts[status] += 1
        else:
            counts["failed"] += 1

        price = result.get("min_weekly_price_aud")
        price_str = f"${price}/wk" if price else "price=?"
        conf  = result.get("confidence", "?")

        append_row(OUTPUT_CSV, {
            "id":                  gym["id"],
            "name":                name,
            "addressState":        state,
            "website":             website,
            "status":              status,
            "min_weekly_price_aud": price if price is not None else "",
            "amenities":           json.dumps(result["amenities"]),
            "member_offers":       json.dumps(result["member_offers"]),
            "confidence":          conf,
            "error":               error or "",
        })

        tag = "OK" if status == "success" else status.upper()
        print(f"{tag}  {price_str}  conf={conf}")

        if i < len(remaining):
            time.sleep(args.delay)

    # Final summary
    total = sum(counts.values())
    print(f"\n── Done ({'all' if not args.limit else args.limit} gyms processed) ──")
    print(f"  success    : {counts['success']:>5}")
    print(f"  parse_error: {counts['parse_error']:>5}")
    print(f"  no_website : {counts['no_website']:>5}")
    print(f"  failed     : {counts['failed']:>5}")
    print(f"  total      : {total:>5}")
    print(f"\nOutput: {OUTPUT_CSV.resolve()}")


if __name__ == "__main__":
    main()
